runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -2941,6 +3007,82 @@ void set_hugepage_info(void)
 }
 
 
+static
+void install_SIGBUS_handler(void)
+{
+  struct sigaction act;
+
+  act.sa_flags = SA_RESTART | SA_SIGINFO;
+  if (sigemptyset(&act.sa_mask) != 0) {
+    CHPL_INTERNAL_ERROR("sigemptyset() failed");
+  }
+  act.sa_sigaction = SIGBUS_handler;
+  if (sigaction(SIGBUS, &act, &previous_SIGBUS_sigact) != 0) {
+    CHPL_INTERNAL_ERROR("sigaction(SIGBUS) failed");
+  }
+}
+
+
+static
+void SIGBUS_handler(int signo, siginfo_t *info, void *context)
+{
+  //
+  // See if this was within an as-yet-unregistered region where we may
+  // be faulting in pages as a side effect of initialization.  If so,
+  // report allocation failure and exit.  Note that we're not holding
+  // the dynamic registration mutex and thus mem_regions.mreg_cnt may
+  // vary while we're executing this loop.  However, if the SIGBUS is
+  // the result of a fault-in failure on some region in the table, that
+  // count won't drop below the index of that region because we can't
+  // remove the region until after it's registered and we can't finish
+  // registering it without faulting in all its pages.
+  // 
+  // Also note that although we replicate the format of the "official"
+  // message chpl_memhook_check_post() produces, we can't just call it
+  // directly or use the same functions as it does, because they're not
+  // signal-safe.
+  //
+  if (info->si_code == BUS_ADRERR) {
+    uint64_t addr = (uint64_t) (intptr_t) info->si_addr;
+
+    int mr_i;
+    mem_region_t* mr;
+
+    for (mr_i = mem_regions.mreg_cnt - 1, mr = &mem_regions.mregs[mr_i];
+         mr_i >= 0 && (addr < mr->addr
+                       || addr >= mr->addr + mrtl_len(mr->len)
+                       || mrtl_isReg(mr->len));
+         mr_i--, mr--)
+      ;
+
+    if (mr_i >= 0) {
+      char buf[1024];
+      chpl_error_vs(buf, sizeof(buf),
+                    mr_mregs_supplement[mr_i].ln, mr_mregs_supplement[mr_i].fn,
+                    "Out of memory allocating \"%s\"",
+                    chpl_mem_descString(mr_mregs_supplement[mr_i].desc));
+      write(fileno(stderr), buf, strlen(buf));
+      exit_without_cleanup = true;
+      atomic_thread_fence(memory_order_release);
+      chpl_exit_any(1);
+    }
+  }
+
+  //
+  // This was not due to allocation failure on a hugepage we were trying
+  // to fault in.  Restore the previous handler and re-raise the SIGBUS.
+  // It should get dealt with as it would have without our handler.  If
+  // the old handler returns and execution continues, install our new
+  // handler again.
+  //
+  if (sigaction(SIGBUS, &previous_SIGBUS_sigact, NULL) != 0) {
+    CHPL_INTERNAL_ERROR("sigaction(SIGBUS) to reinstall old handler failed");
+  }
+  raise(SIGBUS);
+  install_SIGBUS_handler();
```

Comment:
Does it matter what happens if different threads are taking ``SIGBUS`` at the same time, some from hugepage fault-ins, and some not?  ``sigaction()`` sets the behavior for the whole process, but more than one thread can be in the handler at the same time.  (Or in different handlers if the signal arrives with a different ``sigaction`` state.)  Maybe it doesn't matter as long as at least one of the handlers will do something fatal.

Reply:
Hmm indeed.  The solution would be to do `pthread_sigmask(SIG_UNBLOCK, <mask_with_just_SIGBUS>, NULL)` before the `raise()` call, right?  (And maybe restore it afterward, just for completeness.)  Without `SA_NODEFER`, while we're handling a `SIGBUS` due to fault-in that signal will be blocked in the process, if we can believe the `sigaction(3p)` man page.  So we won't get another `SIGBUS` for a hugepage fault-in in this thread while we have the previous handler re-installed in place of our new one.  Then we're safe just changing the sigmask for our thread and trying the old handler, then putting the sigmask and new handler back if the `raise()` returns.